File: 2023/day25/php/solution.php
<?php
/**
 * Advent of Code 2023 - Day 25: Snowverload
 * Find the minimum cut of 3 edges that divides the graph into two components.
 *
 * Uses edge betweenness centrality: edges that form the cut between two large
 * components will have high betweenness (many shortest paths pass through them).
 */

function parseInput($filename) {
    $graph = [];
    $lines = file($filename, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);

    foreach ($lines as $line) {
        $parts = explode(': ', $line);
        $left = trim($parts[0]);
        $neighbors = explode(' ', trim($parts[1]));

        if (!isset($graph[$left])) {
            $graph[$left] = [];
        }

        foreach ($neighbors as $neighbor) {
            $graph[$left][$neighbor] = true;
            if (!isset($graph[$neighbor])) {
                $graph[$neighbor] = [];
            }
            $graph[$neighbor][$left] = true;
        }
    }

    return $graph;
}

function bfsComponentSize($graph, $start, $excludedEdges) {
    $visited = [$start => true];
    $queue = [$start];
    $head = 0;

    while ($head < count($queue)) {
        $node = $queue[$head++];

        foreach (array_keys($graph[$node]) as $neighbor) {
            $edge = [$node, $neighbor];
            sort($edge);
            $edgeKey = $edge[0] . '-' . $edge[1];

            if (!isset($visited[$neighbor]) && !isset($excludedEdges[$edgeKey])) {
                $visited[$neighbor] = true;
                $queue[] = $neighbor;
            }
        }
    }

    return count($visited);
}

function computeEdgeBetweenness($graph, $sampleNodes = null) {
    $edgeCount = [];
    $nodes = array_keys($graph);

    // Sample nodes for efficiency
    if ($sampleNodes !== null && count($nodes) > $sampleNodes) {
        mt_srand(42);
        shuffle($nodes);
        $nodes = array_slice($nodes, 0, $sampleNodes);
    }

    foreach ($nodes as $source) {
        // BFS to find shortest paths
        $dist = [$source => 0];
        $pred = [];
        $queue = [$source];
        $head = 0;

        while ($head < count($queue)) {
            $node = $queue[$head++];

            foreach (array_keys($graph[$node]) as $neighbor) {
                if (!isset($dist[$neighbor])) {
                    $dist[$neighbor] = $dist[$node] + 1;
                    $pred[$neighbor] = [$node];
                    $queue[] = $neighbor;
                } elseif ($dist[$neighbor] == $dist[$node] + 1) {
                    $pred[$neighbor][] = $node;
                }
            }
        }

        // Number of shortest paths to each node
        $numPaths = [];
        $numPaths[$source] = 1.0;

        // Sort nodes by distance
        $sortedNodes = array_keys($dist);
        usort($sortedNodes, function($a, $b) use ($dist) {
            return $dist[$a] <=> $dist[$b];
        });

        foreach ($sortedNodes as $node) {
            if (!isset($numPaths[$node])) {
                $numPaths[$node] = 0.0;
            }
            if (isset($pred[$node])) {
                foreach ($pred[$node] as $p) {
                    $numPaths[$node] += $numPaths[$p];
                }
            }
        }

        // Accumulate edge betweenness (reverse BFS order)
        $dependency = [];

        // Sort nodes by distance (descending)
        usort($sortedNodes, function($a, $b) use ($dist) {
            return $dist[$b] <=> $dist[$a];
        });

        foreach ($sortedNodes as $node) {
            if (!isset($dependency[$node])) {
                $dependency[$node] = 0.0;
            }

            if (isset($pred[$node])) {
                foreach ($pred[$node] as $p) {
                    $edge = [$p, $node];
                    sort($edge);
                    $edgeKey = $edge[0] . '-' . $edge[1];

                    // Weight by path fraction
                    $frac = $numPaths[$p] / $numPaths[$node];
                    $contrib = $frac * (1 + $dependency[$node]);

                    if (!isset($edgeCount[$edgeKey])) {
                        $edgeCount[$edgeKey] = 0.0;
                    }
                    $edgeCount[$edgeKey] += $contrib;

                    if (!isset($dependency[$p])) {
                        $dependency[$p] = 0.0;
                    }
                    $dependency[$p] += $contrib;
                }
            }
        }
    }

    return $edgeCount;
}

function findCutEdges($graph) {
    // Compute edge betweenness with sampling for speed
    $edgeBetweenness = computeEdgeBetweenness($graph, 100);

    // Sort edges by betweenness (highest first)
    arsort($edgeBetweenness);

    $totalNodes = count($graph);

    // Try removing top candidate edges
    // We need to find 3 edges that disconnect the graph
    $topEdges = array_slice(array_keys($edgeBetweenness), 0, 20);

    for ($i = 0; $i < count($topEdges); $i++) {
        for ($j = $i + 1; $j < count($topEdges); $j++) {
            for ($k = $j + 1; $k < count($topEdges); $k++) {
                $excluded = [
                    $topEdges[$i] => true,
                    $topEdges[$j] => true,
                    $topEdges[$k] => true
                ];

                $start = array_key_first($graph);
                $size1 = bfsComponentSize($graph, $start, $excluded);

                if ($size1 < $totalNodes) {
                    // Graph is disconnected!
                    $size2 = $totalNodes - $size1;
                    return $size1 * $size2;
                }
            }
        }
    }

    return null;
}

function part1($filename) {
    $graph = parseInput($filename);
    return findCutEdges($graph);
}

function part2($filename) {
    return "Push the big red button!";
}

// Main execution
$inputFile = __DIR__ . '/../input.txt';

echo "Part 1: " . part1($inputFile) . "\n";
echo "Part 2: " . part2($inputFile) . "\n";

File: 2023/day25/java/Solution.java
import java.io.*;
import java.util.*;

/**
 * Advent of Code 2023 - Day 25: Snowverload
 * Find the minimum cut of 3 edges that divides the graph into two components.
 *
 * Uses edge betweenness centrality: edges that form the cut between two large
 * components will have high betweenness (many shortest paths pass through them).
 */
public class Solution {

    static class Edge {
        String a, b;

        Edge(String node1, String node2) {
            // Keep edges sorted for consistent comparison
            if (node1.compareTo(node2) < 0) {
                a = node1;
                b = node2;
            } else {
                a = node1;
                b = node2;
            }
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (!(o instanceof Edge)) return false;
            Edge edge = (Edge) o;
            return a.equals(edge.a) && b.equals(edge.b);
        }

        @Override
        public int hashCode() {
            return Objects.hash(a, b);
        }

        @Override
        public String toString() {
            return a + "-" + b;
        }
    }

    /**
     * Parse the input file into an adjacency list representation.
     */
    static Map<String, Set<String>> parseInput(String filename) throws IOException {
        Map<String, Set<String>> graph = new HashMap<>();

        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = br.readLine()) != null) {
                line = line.trim();
                if (line.isEmpty()) continue;

                String[] parts = line.split(": ");
                String left = parts[0];
                String[] neighbors = parts[1].split(" ");

                graph.putIfAbsent(left, new HashSet<>());
                for (String neighbor : neighbors) {
                    graph.get(left).add(neighbor);
                    graph.putIfAbsent(neighbor, new HashSet<>());
                    graph.get(neighbor).add(left);
                }
            }
        }

        return graph;
    }

    /**
     * BFS to find component size, ignoring excluded edges.
     */
    static int bfsComponentSize(Map<String, Set<String>> graph, String start, Set<Edge> excludedEdges) {
        Set<String> visited = new HashSet<>();
        Queue<String> queue = new LinkedList<>();

        visited.add(start);
        queue.offer(start);

        while (!queue.isEmpty()) {
            String node = queue.poll();
            Set<String> neighbors = graph.get(node);
            if (neighbors == null) continue;

            for (String neighbor : neighbors) {
                Edge edge = new Edge(node, neighbor);
                if (!visited.contains(neighbor) && !excludedEdges.contains(edge)) {
                    visited.add(neighbor);
                    queue.offer(neighbor);
                }
            }
        }

        return visited.size();
    }

    /**
     * Compute approximate edge betweenness centrality.
     * Higher values indicate edges that many shortest paths pass through.
     */
    static Map<Edge, Double> computeEdgeBetweenness(Map<String, Set<String>> graph, int sampleNodes) {
        Map<Edge, Double> edgeCount = new HashMap<>();
        List<String> nodes = new ArrayList<>(graph.keySet());

        // Sample nodes for efficiency
        if (sampleNodes > 0 && nodes.size() > sampleNodes) {
            Random rand = new Random(42);
            Collections.shuffle(nodes, rand);
            nodes = nodes.subList(0, sampleNodes);
        }

        for (String source : nodes) {
            // BFS to find shortest paths
            Map<String, Integer> dist = new HashMap<>();
            Map<String, List<String>> pred = new HashMap<>();
            Queue<String> queue = new LinkedList<>();

            dist.put(source, 0);
            queue.offer(source);

            while (!queue.isEmpty()) {
                String node = queue.poll();
                int nodeDist = dist.get(node);

                Set<String> neighbors = graph.get(node);
                if (neighbors == null) continue;

                for (String neighbor : neighbors) {
                    if (!dist.containsKey(neighbor)) {
                        dist.put(neighbor, nodeDist + 1);
                        pred.putIfAbsent(neighbor, new ArrayList<>());
                        pred.get(neighbor).add(node);
                        queue.offer(neighbor);
                    } else if (dist.get(neighbor) == nodeDist + 1) {
                        pred.putIfAbsent(neighbor, new ArrayList<>());
                        pred.get(neighbor).add(node);
                    }
                }
            }

            // Backtrack to count edge usage
            // Number of shortest paths to each node
            Map<String, Double> numPaths = new HashMap<>();
            numPaths.put(source, 1.0);

            List<String> sortedNodes = new ArrayList<>(dist.keySet());
            sortedNodes.sort(Comparator.comparingInt(dist::get));

            for (String node : sortedNodes) {
                List<String> predecessors = pred.get(node);
                if (predecessors != null) {
                    for (String p : predecessors) {
                        double prevPaths = numPaths.getOrDefault(p, 0.0);
                        numPaths.put(node, numPaths.getOrDefault(node, 0.0) + prevPaths);
                    }
                }
            }

            // Accumulate edge betweenness (reverse BFS order)
            Map<String, Double> dependency = new HashMap<>();
            sortedNodes.sort((a, b) -> Integer.compare(dist.get(b), dist.get(a)));

            for (String node : sortedNodes) {
                List<String> predecessors = pred.get(node);
                if (predecessors != null) {
                    for (String p : predecessors) {
                        Edge edge = new Edge(p, node);
                        double frac = numPaths.get(p) / numPaths.get(node);
                        double contrib = frac * (1.0 + dependency.getOrDefault(node, 0.0));
                        edgeCount.put(edge, edgeCount.getOrDefault(edge, 0.0) + contrib);
                        dependency.put(p, dependency.getOrDefault(p, 0.0) + contrib);
                    }
                }
            }
        }

        return edgeCount;
    }

    /**
     * Find the 3 edges to cut using edge betweenness.
     */
    static Integer findCutEdges(Map<String, Set<String>> graph) {
        // Compute edge betweenness with sampling for speed
        Map<Edge, Double> edgeBetweenness = computeEdgeBetweenness(graph, 100);

        // Sort edges by betweenness (highest first)
        List<Edge> sortedEdges = new ArrayList<>(edgeBetweenness.keySet());
        sortedEdges.sort((e1, e2) -> Double.compare(edgeBetweenness.get(e2), edgeBetweenness.get(e1)));

        int totalNodes = graph.size();

        // Try removing top candidate edges
        // We need to find 3 edges that disconnect the graph
        List<Edge> topEdges = sortedEdges.subList(0, Math.min(20, sortedEdges.size()));

        for (int i = 0; i < topEdges.size(); i++) {
            for (int j = i + 1; j < topEdges.size(); j++) {
                for (int k = j + 1; k < topEdges.size(); k++) {
                    Set<Edge> excluded = new HashSet<>();
                    excluded.add(topEdges.get(i));
                    excluded.add(topEdges.get(j));
                    excluded.add(topEdges.get(k));

                    String start = graph.keySet().iterator().next();
                    int size1 = bfsComponentSize(graph, start, excluded);

                    if (size1 < totalNodes) {
                        // Graph is disconnected!
                        int size2 = totalNodes - size1;
                        return size1 * size2;
                    }
                }
            }
        }

        return null;
    }

    /**
     * Solve Part 1: Find the 3-edge cut and return product of component sizes.
     */
    static int part1(String filename) throws IOException {
        Map<String, Set<String>> graph = parseInput(filename);
        Integer result = findCutEdges(graph);
        return result != null ? result : 0;
    }

    /**
     * Part 2: Day 25 Part 2 is traditionally unlocked by having 49 stars.
     * There's no computation needed - just push the button!
     */
    static String part2(String filename) {
        return "Push the big red button!";
    }

    public static void main(String[] args) throws IOException {
        String inputFile = "../input.txt";

        System.out.println("Part 1: " + part1(inputFile));
        System.out.println("Part 2: " + part2(inputFile));
    }
}
